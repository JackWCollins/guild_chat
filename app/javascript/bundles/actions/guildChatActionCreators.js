/* eslint-disable import/prefer-default-export */

export const USER_LOGIN = 'USER_LOGIN';
export const login = (userId, userName) => ({
  type: USER_LOGIN,
  userId,
  userName
});

export const USER_LOGOUT = 'USER_LOGOUT';
export const logout = () => ({
  type: USER_LOGOUT,
});

export const LOAD_CONVERSATIONS_REQUEST = 'LOAD_CONVERSATIONS_REQUEST';
const requestConversations = () => ({
  type: LOAD_CONVERSATIONS_REQUEST
});

export const LOAD_CONVERSATIONS_SUCCESS = 'LOAD_CONVERSATIONS_SUCCESS';
const receiveConversations = (conversations) => ({
  type: LOAD_CONVERSATIONS_SUCCESS,
  conversations
});

export function loadConversations(userId) {
  return function (dispatch) {
    dispatch(requestConversations());

    return fetch('/users/'+userId+'/conversations', {
      method: 'GET',
      headers: new Headers({
        'Content-Type': 'application/json'
      })
    })
      .then(
        response => response.json(),
        error => console.log('An error occurred while loading user conversations: ', error)
      )
      .then(json => dispatch(receiveConversations(json)));
  };
}

export const SEND_MESSAGE_REQUEST = 'SEND_MESSAGE_REQUEST';
const sendMessageRequest = () => ({
  type: SEND_MESSAGE_REQUEST
});

export const SEND_MESSAGE_SUCCESS = 'SEND_MESSAGE_SUCCESS';
const sendMessageSuccess = (message) => ({
  type: SEND_MESSAGE_SUCCESS,
  message
});

export function sendMessage(userId, conversationId, messageBody) {
  console.log("sendMessage action: ", userId, conversationId, messageBody)
  return function (dispatch) {
    dispatch(sendMessageRequest());

    return fetch('/messages', {
      method: 'POST',
      body: JSON.stringify({user_id: userId, conversation_id: conversationId, body: messageBody}),
      headers: new Headers({
        'Content-Type': 'application/json'
      })
    })
      .then(
        response => response.json(),
        error => console.log('An error occurred while sending the message: ', error)
      )
      .then(json => dispatch(sendMessageSuccess(json)))
  }
}

export const LOAD_CONVERSATION_REQUEST = 'LOAD_CONVERSATION_REQUEST';
const loadConversationRequest = () => ({
  type: LOAD_CONVERSATION_REQUEST
});

export const LOAD_CONVERSATION_SUCCESS = 'LOAD_CONVERSATION_SUCCESS';
const loadConversationSuccess = (conversation) => ({
  type: LOAD_CONVERSATION_SUCCESS,
  conversation
});

export function selectConversation(conversationId) {
  return function (dispatch) {
    dispatch(loadConversationRequest());

    return fetch('/conversations/'+conversationId, {
      method: 'GET',
      headers: new Headers({
        'Content-Type': 'application/json'
      })
    })
      .then(
        response => response.json(),
        error => console.log('An error occurred while loading the conversation: ', error)
      )
      .then(json => dispatch(loadConversationSuccess(json)))
  }
}
