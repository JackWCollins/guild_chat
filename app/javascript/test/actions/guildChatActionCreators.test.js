import * as actions from '../../bundles/actions/guildChatActionCreators';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('actions', () => {
  it('login should return the userId and userName', () => {
    expect(actions.login(123, "Jack")).toMatchSnapshot();
  });

  // This is testing an action that is using `thunk`.
  // Test for both actions when the mock HTTP request is returned.
  it('loadConversations should load the user conversations', () => {

    fetch.mockResponse(JSON.stringify(
      [
        { id: 1,
          messages: [
            {id: 10, body: "Cool!", user: {id: 4, first_name: "Jack", last_name: "Collins"}}
          ],
          users: [
            {id: 4, first_name: "Jack", last_name: "Collins"},
            {id: 5, first_name: "Annie", last_name: "Collins"}
          ]
        }
      ]
    ));

    const expectedActions = [
      { type: 'LOAD_CONVERSATIONS_REQUEST' },
      { type: 'LOAD_CONVERSATIONS_SUCCESS', conversations:
        [
          { id: 1,
            messages: [
              {id: 10, body: "Cool!", user: {id: 4, first_name: "Jack", last_name: "Collins"}}
            ],
            users: [
              {id: 4, first_name: "Jack", last_name: "Collins"},
              {id: 5, first_name: "Annie", last_name: "Collins"}
            ]
          }
        ]
      }
    ];
    const store = mockStore({ home: { conversations: [] } });

    return (
      store.dispatch(actions.loadConversations(4))
        .then( () => {
          expect(store.getActions()).toEqual(expectedActions);
        })
    );
  });
});
