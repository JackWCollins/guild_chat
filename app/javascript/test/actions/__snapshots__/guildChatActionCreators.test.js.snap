// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions login should return the userId and userName 1`] = `
Object {
  "type": "USER_LOGIN",
  "userId": 123,
  "userName": "Jack",
}
`;
